tree_wide: add metrics for ops dashboard

VladLazar opened the pull request:
## Cover letter

A number of metrics are added in this PR to get a set of dashboards designed by @tmgstevens
working with public_metrics. The list is available in the release notes.

<!-- Use the GitHub keyword `Fixes` to link to bug(s) this PR will fix. -->
Fixes #7059 
## Backport Required

<!-- Specify which branches this should be backported to, e.g.: -->
- [X] not a bug fix
- [ ] issue does not exist in previous branches
- [ ] papercut/not impactful enough to backport
- [ ] v22.3.x
- [ ] v22.2.x
- [ ] v22.1.x

## Release notes
A number of new metrics have been introduced. Full list below.

redpanda_raft_leadership_changes:
* Description: Number of leadership changes accross all partitions of a
  given topic.
* Labels: redpanda_namespace, redpanda_topic
* Type: counter

redpanda_reactor_utilization:
* Description: Reactor utilization percentage
* Labels: shard
* Type: gauge

redpanda_application_build:
* Description: Redpanda build information
* Labels: redpanda_version, redpanda_revision
* Type: gauge

redpanda_rpc_active_connections:
* Description: Count of currently active connections
* Labels: redpanda_server (e.g. kafka)
* Type: gauge

VladLazar commented:
CI failure is https://github.com/redpanda-data/redpanda/issues/7043

jcsp commented on src/v/redpanda/application.cc:
```diff
@@ -411,7 +411,8 @@ void application::setup_public_metrics() {
        {"memory_allocated_memory", ssx::metrics::public_metrics_handle},
-       {"memory_free_memory", ssx::metrics::public_metrics_handle}})
+       {"memory_free_memory", ssx::metrics::public_metrics_handle},
+       {"reactor_utilization", ssx::metrics::public_metrics_handle}})
```
Hmm.  I think anything with `reactor` in the name is definitely too implementation-detailed for the public metrics.

Internally to seastar it looks like this one is a decaying load statistic, whereas our currently redpanda_cpu_busy_seconds_total metric is a giving a direct measurement of how fast the process is accumulating busy time (i.e. take the gradient and you get how busy the CPU was in your sample period).

So two questions:
- A) do we need this?
- B) how do we explain to the user which statistic is the right one for a dashboard that wants to show "how busy is my CPU" in one chart?

VladLazar replied:
I agree that the name is too specific, but the metrics API doesn't allow renaming at the moment (maybe it should; our seastar metrics implementation has diverged quite a bit from upstream).

Done a bit of research and `reactor_utilization` seems to be the most accurate metric of how much work is going on at any given point. For some reason I didn't think of using the rate of change in `redpanda_cpu_busy_seconds_total` to estimate this. Let me chat to Tristan and see if that would work.

VladLazar replied:
Endede up removing it. We'll got with the rate of `redpanda_cpu_busy_seconds_total`.

jcsp commented:
LGTM apart from the reactor utilization metric, where we need a clear story about what it is for and when users should use the existing CPU metric vs this one

VladLazar commented:
Change in [force-push](https://github.com/redpanda-data/redpanda/compare/32274ebe1487f35b7dcd873ebe232036a8e81463..b77687053a2cb4f30bc5ab11c8602b9efa1263db):
* Dropped `reactor_utilization` metric.
* Dropped commit fixing recovery throttle metric name as that was fixed already.

BenPope commented on src/v/redpanda/application.cc:
```diff
@@ -416,10 +416,16 @@ void application::setup_public_metrics() {
     _public_metrics.start().get();
 
+    const auto version_label = ssx::metrics::make_namespaced_label("version")(
+      redpanda_git_version());
+    const auto revision_label = ssx::metrics::make_namespaced_label("revision")(
+      redpanda_git_revision());
+    const auto build_labels = {version_label, revision_label};
+
     _public_metrics
       .invoke_on(
         ss::this_shard_id(),
-        [](auto& public_metrics) {
+        [build_labels](auto& public_metrics) {
```
very minor nit (scope, copies, LoC, etc.):
```suggestion
        [](auto& public_metrics) {
            const auto build_labels = {
              ssx::metrics::make_namespaced_label("version")(
                redpanda_git_version()),
              ssx::metrics::make_namespaced_label("revision")(
                redpanda_git_revision())};

```

BenPope commented:
I think you might like to tick the backport button if it's destined for `v22.3.x`

VladLazar commented:
/backport v22.3.x
